ops/terraform/modules/resources/s3/main.tf

```diff
@@ -19,22 +19,42 @@ resource "aws_s3_bucket" "main" {
   acl                     = var.acl
   tags                    = local.tags
 
+  # Always apply encryption, Customer CMK or AWS AES256
   server_side_encryption_configuration {
     rule {
-      apply_server_side_encryption_by_default {
-        kms_master_key_id = var.kms_key_id
-        sse_algorithm     = "aws:kms"
+      dynamic "apply_server_side_encryption_by_default" {
+        for_each = var.kms_key_id == null ? ["doit"] : []
+        content {
+          sse_algorithm     = "AES256"
+        }
+      }
+      dynamic "apply_server_side_encryption_by_default" {
+        for_each = var.kms_key_id != null ? ["doit"] : []
+        content {
+          sse_algorithm     = "aws:kms"
+          kms_master_key_id = var.kms_key_id
+        }
```

Comment:
I did some testing on this and I think you might be able to get away with just setting `kms_master_key_id` to null conditionally when `sse_algorithm` is AES256 rather than generating the whole property dynamically.

Reply:
Good point. Will try it out.